### PR TITLE
Update 'updated_at' when enabled in replace()

### LINKED ISSUE #4624: Bug: $model->replace() not generated timestamp
**Describe the bug**
When using $model->replace($data), the timestamp field not generating 

**CodeIgniter 4 version**
Codeigniter 4.1.1

**Context**
 - OS: Debian Linux
 - Web server Nginx
 - PHP version 7.4

### PR DESCRIPTION
**Pull request by paulbalandan**
**Description**
Fixes #4624 
Supersedes and closes #4662 

**Checklist:**
- [x] Securely signed commits

### CODE REVIEW

**Comment by MGatner** [comment]
What's happening with this? Why are we all failing at this issue... 🤦‍♂️

**Comment by paulbalandan** [comment]
The Postgre builder actually has a bug which I discovered that testMode will make replace still execute the SQL. For Sqlsrv, I am not sure where went wrong. I am still to setup sql server locally so I can see this thru xdebug.

**Comment by MGatner** [comment]
😭

**Review by MGatner — APPROVED** [review]
I don't understand what happened there but it's working now and I'm happy.

**Review by sfadschm — COMMENTED** [review]
Is the check for `$data` still needed?

**Comment by paulbalandan** [comment]
> Is the check for `$data` in the `if` statement still needed?

You mean the validation part? I think so because replace is like update, isn't it? Why do you think it's not needed anymore?

File: system/BaseModel.php [hunk 1]
```diff
@@ -1090,12 +1090,19 @@ public function onlyDeleted()
      */
     public function replace(array $data = null, bool $returnSQL = false)
     {
+        $data = $data ?? [];
```
Comment (sfadschm) [inline system/BaseModel.php:1093]: Sorry, I messed up the reference on mobile...
I was thinking about this line. I do not really understand why it is needed here.

If I'm correct, the null coalescing operator on `$data` will only reset it to `[]` if `$data` is `null``, as it is typehinted.
In line 1096, we do `if($data && ...  && validate($data))`, both `[]` and `null` would evaluate `false` in the first condition, so we do not need to enure `$data` to be array here.

In the second `if` block, `array_key_exists`  requires an array, so I think we can pull the null coalescence directly before the second `if`. It is not needed in the first `if`.

I hope you can understand what I mean 😆

Comment (paulbalandan) [reply system/BaseModel.php:1093]: Ah yes. It's misplaced. I'll fix that when I'm not on mobile. Thanks!

**Review by MGatner — APPROVED** [review]

**Pull request merged**